Sources/NIO/SocketOptionProvider.swift

```diff
@@ -56,7 +74,23 @@ public protocol SocketOptionProvider {
     ///     - value: The value to set the socket option to.
     /// - returns: An `EventLoopFuture` that fires when the option has been set,
     ///     or if an error has occurred.
-    func unsafeSetSocketOption<Value>(level: SocketOptionLevel, name: SocketOptionName, value: Value) -> EventLoopFuture<Void>
+    func unsafeSetSocketOption<Value>(level: NIOBSDSocket.OptionLevel, name: NIOBSDSocket.Option, value: Value) -> EventLoopFuture<Void>
+
+    #if !os(Windows)
+        /// Obtain the value of the socket option for the given level and name.
+        ///
+        /// This function is not memory-safe: if you set the generic type parameter incorrectly,
+        /// this function will still execute, and this can cause you to incorrectly interpret memory
+        /// and thereby read uninitialized or invalid memory. If at all possible, please use one of
+        /// the safe functions defined by this protocol.
+        ///
+        /// - parameters:
+        ///     - level: The socket option level, e.g. `SOL_SOCKET` or `IPPROTO_IP`.
+        ///     - name: The name of the socket option, e.g. `SO_REUSEADDR`.
+        /// - returns: An `EventLoopFuture` containing the value of the socket option, or
+        ///     any error that occurred while retrieving the socket option.
+        func unsafeGetSocketOption<Value>(level: SocketOptionLevel, name: SocketOptionName) -> EventLoopFuture<Value>
+    #endif
```

Comment:
We can't add new protocol requirements without having default implementations. Mind providing a default implementation here?

Reply:
I thought I had added that, did I miss something?

Reply:
https://github.com/apple/swift-nio/pull/1461/files#diff-0472f0322d1f2b7085ece7bd3f5ed3d5R271-R275 and https://github.com/apple/swift-nio/pull/1461/files#diff-0472f0322d1f2b7085ece7bd3f5ed3d5R292-R294 provide the default implementation (or am I misunderstanding something?)

Reply:
I see implementations on `SocketChannel`, but I mean there must be a default implementation in a protocol extension. Otherwise types that previously conformed now will not.

Reply:
Oh, different types >.< ... I get it now.